prov/sockets/src/sock_fabric.c

```diff
@@ -658,35 +702,45 @@ static int sock_getinfo(uint32_t version, const char *node, const char *service,
 		return ret;
 
 	slist_init(&addr_list);
-	if (!node  && (!hints || !hints->src_addr || (!(flags & FI_SOURCE) &&
-	     !hints->dest_addr))) {
-		addr_list = sock_get_list_of_addr();
+	/* Returns loopback address if no other interfaces are available */
+	sock_get_list_of_addr(&addr_list);
+
+	ret = 1;
+	if (flags & FI_SOURCE) {
+		if (node)
+			ret = sock_node_matches_interface(addr_list, node);
+	} else if (hints && hints->src_addr) {
+		ret = sock_addr_matches_interface(addr_list, (struct sockaddr_in *)hints->src_addr);
+	}
+	if (!ret) {
+		SOCK_LOG_ERROR("Couldn't find a match with local interfaces\n");
+		return -FI_ENODATA;
 	}
 
 	*info = tail = NULL;
-	if (slist_empty(&addr_list)) {
+	if (node ||
+	     (!(flags & FI_SOURCE) && hints && hints->src_addr) ||
+	     (!(flags & FI_SOURCE) && hints && hints->dest_addr)) {
+		sock_free_addr_list(&addr_list);
 		return sock_node_getinfo(node, service, flags, hints, info, &tail);
-	} else {
-		while (!slist_empty(&addr_list)) {
-			entry = slist_remove_head(&addr_list);
-			host_entry = container_of(entry, struct sock_host_list_entry, entry);
-			node = host_entry->hostname;
-			ret = sock_node_getinfo(node, service, flags, hints, info, &tail);
-			free(host_entry);
-			if (ret) {
-				if (ret == -FI_ENODATA)
-					continue;
-				goto out;
-			}
-		}
 	}
-	return 0;
-out:
+
 	while (!slist_empty(&addr_list)) {
 		entry = slist_remove_head(&addr_list);
 		host_entry = container_of(entry, struct sock_host_list_entry, entry);
+		node = host_entry->hostname;
+		flags |= FI_SOURCE;
+		ret = sock_node_getinfo(node, service, flags, hints, info, &tail);
 		free(host_entry);
```

Comment:
Ok, a cleaner way is to just free the entries before exiting the loop, and then do a break. That can avoid the goto out.